### ISSUE TITLE
Parsing issues

### ISSUE THREAD

**Issue by wichtounet** [SonarOpenCommunity/sonar-cxx#733]
Hi, 

I have some weird parsing issues with some files of mine. It seems the problem occurs when I use my SFINAE macro in templates

```
namespace etl {

//TODO Fix the false positives in sonar-cxx for these functions

/*!
 * \brief Builds an expression representing the subtraction of lhs and rhs
 * \param lhs The left hand side expression
 * \param rhs The right hand side expression
 * \return An expression representing the subtraction of lhs and rhs
 */
template <typename LE, typename RE, cpp_enable_if(is_etl_expr<LE>::value, is_etl_expr<RE>::value)>
auto operator-(LE&& lhs, RE&& rhs) -> detail::left_binary_helper<LE, RE, minus_binary_op> {
    validate_expression(lhs, rhs);

    return {lhs, rhs};
}

...
```

I get a syntax error on the "template " line with : syntax error, skip 'LE'
And, probably not related, but weird, I get a syntax error on the namespace line: syntax error, skip 'etl'

The sources for the cpp_enable_if macro are found, but I don't include the sources for std::enable_if (because If I do, it is much slower and I get much more errors). 

In itself, it is not a big issue, but it causes a lot of false positive, in this case "Undocumented API: operator-" and in other cases the body of the function causes false positives too. 

Do you have an idea where could that come from ?

**Comment by guwirth** [comment]
I would play around a bit with the code sample with the sslr-cxx-toolkit, e.g.
* template outside namespace
* remove body of the template
* remove cpp_enable_if
...

To see what cause the problem...

**Comment by guwirth** [comment]
> The sources for the cpp_enable_if macro are found, but I don't include the sources for std::enable_if (because If I do, it is much slower and I get much more errors).

Just another hint for your understanding. The plugin checks the code only against the grammar (syntax check only) without any semantic check. This means includes are only needed to resolve syntax issues. But in most cases defines to mock things away are sufficient (https://github.com/SonarOpenCommunity/sonar-cxx/wiki/Dealing-with-compiler-specific-code-pieces).

**Comment by wichtounet** [comment]
The only thing that removes the problem is to remove the cpp_enable_if completely. 

Even if I do the macro expansion by hand, it still has issues. 

If I simplify up to this: 

    template <typename LE, typename RE, typename std::enable_if<is_etl_expr<LE>::value, int>::type = 42>
    void operator-(LE&& lhs, RE&& rhs){}

It doesn't work, but if I simplify with enable_if_t:

    template <typename LE, typename RE, std::enable_if_t<is_etl_expr<LE>::value, int> = 42>
    void operator-(LE&& lhs, RE&& rhs){}

it works...

It may come from the fact that the typename keyword in the first example is used not to introduce a new template parameter but for the enable_if.

**Comment by guwirth** [comment]
problem not reproducible with this:
```C++
namespace etl {
   template <typename LE, typename RE, cpp_enable_if(is_etl_expr<LE>::value, is_etl_expr<RE>::value)>
   auto operator-(LE&& lhs, RE&& rhs) -> detail::left_binary_helper<LE, RE, minus_binary_op> {
      validate_expression(lhs, rhs);
      return {lhs, rhs};
   }
}
```
problem reproducible with this:
```C++
template <typename LE, typename RE, typename std::enable_if<is_etl_expr<LE>::value, int>::type = 42>
void operator-(LE&& lhs, RE&& rhs){}
```
problem is `typename` before `std::enable_if`, this is working:
```C++
template <typename LE, typename RE, std::enable_if<is_etl_expr<LE>::value, int>::type = 42>
void operator-(LE&& lhs, RE&& rhs){}
```
Has to be verified if erroneous case is valid C++.
@wichtounet expect clang is supporting it?

**Comment by wichtounet** [comment]
@guwirth 

> Has to be verified if erroneous case is valid C++.
> @wichtounet expect clang is supporting it?

I've been using this code of time for a long time and it compiles with both g++ 4.9, clang 3.5, clang 3.7 and icc (don't remember the version). 

I don't see why it wouldn't be legal C++. The typename is necessary because the std::enable_if is type-dependent on the template parameter. The int type means it is a non-type template parameter. And template parameters can be unnamed.

**Comment by guwirth** [comment]
@wichtounet: I'm not a template guru, could you explain what the code should do?

```C++
template <typename LE, typename RE, typename std::enable_if<is_etl_expr<LE>::value, int>::type = 42>
void operator-(LE&& lhs, RE&& rhs){}
```
Looking here http://en.cppreference.com/w/cpp/types/enable_if they have a `=` inbetween?

**Comment by wichtounet** [comment]
@guwirth This code forces a substitution failure when the the operator- is used with LE not being an ETL expression (is_etl_expr is a simple traits of mine). The "typename std::enable_if<is_etl_expr<LE>::value, int>::type" is evaluated to int when it is valid, otherwise it is a substitution error (not a failure), making it not consider this function in overload resolution. The " = 42" simply sets a default value. 

As shown in the page you have given, there are many ways of doing SFINAE. You are talking about the case "5", right ? "typename = std::enable_if_t<std::is_array<T>::value>" Here, the typename is used to declare a new template parameter (a type template parameter) with no name and a default value that is set to the evaluation of enable_if_t (subsitution failure or void).  

The problem with this technique is that cannot declare two functions that differs only in there default parameters. Using the technique I use with non-type template parameters, the SFINAE code is part of the template signature and therefore valid to declare several. 

The case I'm using is the case "4": 

    template<class T,
         typename std::enable_if<
             !std::is_trivially_destructible<T>{} &&
             (std::is_class<T>{} || std::is_union<T>{})
            >::type* = nullptr>

This is the same explanation except that instead of int and 42, we use void* and nullptr. 

For me this is the preferred way of doing SFINAE since it does not pollute the return type neither the function parameters. 

I hope it is more clear ?

**Comment by guwirth** [comment]
Hi @wichtounet thanks a lot for the template training :-)

Ok what I understood, this is
```C++
template <typename LE, typename RE, typename std::enable_if<is_etl_expr<LE>::value, int>::type = 42>

if is_etl_expr<LE>::value == true
   template <typename LE, typename RE, typename int>
```
Interesting is the part
```C++
std::enable_if<...>::type
```
Where `type` returns a `type-specifier`. I think our parser can't detect this and think it's `typename parameter-declarartion` which is not valid:

```
template-declaration:
   template < template-parameter-list > declaration

template-parameter-list:
   template-parameter
   template-parameter-list , template-parameter

template-parameter:
   type-parameter
   parameter-declaration

type-parameter:
   type-parameter-key ...(opt) identifier(opt)
   type-parameter-key identifier(opt)= type-id
   template < template-parameter-list > type-parameter-key ...(opt) identifier(opt)
   template < template-parameter-list > type-parameter-key identifier(opt)= id-expression

type-parameter-key:
   class
   typename
```

**Comment by wichtounet** [comment]
@guwirth You're very welcome ;)

Looking at your grammar for "template-parameter", should no this be handled by "parameter-declaration" ?

**Comment by guwirth** [comment]
@wichtounet no. `template-parameter` could be `type-parameter` or `parameter-declaration`. The `::type` is changing the `parameter-declaration` to a `type-parameter`. Only `type-parameter` can have a `type-parameter-key` (typename).

Means grammar has to be extended somehow, maybe 
```
template-parameter:
   type-parameter
   type-parameter-key(opt) parameter-declaration
```
or
```
type-parameter:
    type-parameter-key declarator-id :: "type"
```
We have to play around and make some samples.

**Comment by wichtounet** [comment]
@guwith But a parameter declaration is allowed to have a ```typename```, for instance:

template<typename T>
void test(typename T::iterator* = nullptr){}

Isn't ```typename T::iterator* = nullptr``` exactly what should be allowed in ```template-parameter``` ? And isn't that already parsed in ```parameter-declaration``` ? I don't know the grammar you're using but maybe trying to match ```parameter-declaration``` first could help ?

There may be some tricky cases though.

**Comment by guwirth** [comment]
Will solve this as port of #743. Is a little bit more work and will take some time.

**Comment by wichtounet** [comment]
:+1:

**Comment by guwirth** [comment]
On #743 is a first but incomplete solution. Use cases above are supported now.
Use case below is still not supported:

```C++
// @todo 
// #4, enabled via a template parameter
template<class T,
         typename std::enable_if<
             !std::is_trivially_destructible<T>{} &&
             (std::is_class<T>{} || std::is_union<T>{})
            >::type* = nullptr>
void destroy(T* t)
{
    std::cout << "destroying non-trivially destructible T\n";
    t->~T();
}
```

**Comment by wichtounet** [comment]
Great job :)

I just don't get your comment:

    // C++ special case to handle ::type (not part of standard)

In my opinion, this is part of the standard, isn't it ?

**Comment by wichtounet** [comment]
Is everything fixed in master ?

**Comment by guwirth** [comment]
Yes, you can give it a try.
